gridscale/resource_gridscale_k8s.go

```diff
@@ -310,6 +265,183 @@ func resourceGridscaleK8s() *schema.Resource {
 	}
 }
 
+/*
+deriveK8sTemplateFromResourceDiff derives the k8s service template
+from given resource diff of instance *schema.ResourceDiff.
+The derivation will base on respective resource's diff created by Terraform.
+*/
+func deriveK8sTemplateFromResourceDiff(client *gsclient.Client, d *schema.ResourceDiff) (*gsclient.PaaSTemplate, error) {
+	derivationTypesRequested := 0
+	derivationType := ""
+
+	versionInterface, isVersionSet := d.GetOk("gsk_version")
+	version := versionInterface.(string)
+	releaseInterface, isReleaseSet := d.GetOk("release")
+	release := releaseInterface.(string)
+
+	if isVersionSet && version != "" {
+		derivationTypesRequested += 1
+		derivationType = "version"
+	}
+
+	if isReleaseSet && release != "" {
+		derivationTypesRequested += 1
+		derivationType = "release"
+	}
+
+	if derivationTypesRequested == 0 {
+		return nil, errors.New("either \"release\" or \"gsk_version\" has to be defined")
+	}
+	if derivationTypesRequested > 1 {
+		return nil, errors.New("\"release\" and \"gsk_version\" are not intended to be set at once.")
+	}
+	switch derivationType {
+	case "version":
+		return deriveK8sTemplateFromGSKVersion(client, version)
+	case "release":
+		return deriveK8sTemplateFromRelease(client, release)
+	}
+	return nil, nil
+}
+
+/*
+deriveK8sTemplateFromResourceData derives the k8s service template
+from given resource data of instance *schema.ResourceData.
+The derivation will base on what gets requested.
+*/
+func deriveK8sTemplateFromResourceData(client *gsclient.Client, d *schema.ResourceData) (*gsclient.PaaSTemplate, error) {
+	derivationTypesRequested := 0
+	derivationType := ""
+
+	versionInterface, isVersionSet := d.GetOk("gsk_version")
+	version := versionInterface.(string)
+	releaseInterface, isReleaseSet := d.GetOk("release")
+	release := releaseInterface.(string)
+
+	if !d.IsNewResource() { // case if update of resource is requested
+		if isVersionSet && d.HasChange("version") {
+			derivationTypesRequested += 1
+			derivationType = "version"
+		}
+
+		if isReleaseSet && d.HasChange("release") {
+			derivationTypesRequested += 1
+			derivationType = "release"
+		}
+	} else { // case if creation of resource is requested
+		if isVersionSet {
+			derivationTypesRequested += 1
+			derivationType = "version"
+		}
+
+		if isReleaseSet {
+			derivationTypesRequested += 1
+			derivationType = "release"
+		}
+
+		if derivationTypesRequested == 0 {
+			return nil, errors.New("either \"release\" or \"gsk_version\" has to be defined")
+		}
+	}
+
+	if derivationTypesRequested > 1 {
+		return nil, errors.New("\"release\" and/or \"gsk_version\" are not intended to be set at once.")
+	}
+	switch derivationType {
+	case "version":
+		return deriveK8sTemplateFromGSKVersion(client, version)
+	case "release":
+		return deriveK8sTemplateFromRelease(client, release)
+	}
+	currentTemplateUUID := d.Get("service_template_uuid").(string)
```

Comment:
Yes. I know it is computet. But when you follow the logic, then you'll understand that that line of code just comes into play in case if udate is requested and not in case if creation is requested. In case if creation would be requested following code would come to play

```
	switch derivationType {
	case "version":
		return deriveK8sTemplateFromGSKVersion(client, version)
	case "release":
		return deriveK8sTemplateFromRelease(client, release)
	}
```
In case of update also the code shown in code snippet above would come to play, but if neither release nor GSK version is given at update or is given, but has not changed from current value, then the template to retirieve will be the one that can be derived from the currently known template UUID. I'm noing there nothing else as you already do [here](https://github.com/gridscale/terraform-provider-gridscale/blob/e56aad4c55902654782027b3c564aa8badc596db/gridscale/resource_gridscale_k8s.go#L647).

Reply:
Yes. But I'll need to fetch the template also at the update anyway to get the final version of the template for processinf as I'll need it for the implementation of the node pools. Anyway `deriveK8sTemplateFromResourceData` needs to get called and in case if release or GSK version is not set or set, but value doesn't differ from the already known value, then instead of returning nil for the expected template pointer instance we return the template that is related to the already known UUID.